profiles/utils.py

```diff
@@ -64,11 +64,20 @@ def set_current_service(service):
     _thread_locals.service = service
 
 
+def get_current_request():
+    return getattr(_thread_locals, "request", None)
+
+
 def get_current_user():
-    request = getattr(_thread_locals, "request", None)
+    request = get_current_request()
     return getattr(request, "user", None) if request else None
 
 
+def get_original_client_ip():
+    request = get_current_request()
+    return request.headers.get("x-forwarded-for") if request else None
```

Comment:
The value of the x-forwarded-for is comma separated list of ip-addresses (if there were many hops). So you need to split it and take the first.

Also in many cases (e.g. when running locally) the x-forwarded-for is not necessarily set. I think you should use REMOTE_ADDR in that case.

Some projects add a middleware that sets the REMOTE_ADDR from the x-forwarded-for if that is set, so that later you can just use REMOTE_ADDR.

Reply:
Another thing is, that the usage of x-forwarded-for should be configurable in settings. Because if someone is running this project in production without any proxies, the x-forwarded-for should not be trusted.

Reply:
> Another thing is, that the usage of x-forwarded-for should be configurable in settings. Because if someone is running this project in production without any proxies, the x-forwarded-for should not be trusted.

The X-Forwarded-For can probably be spoofed whether there are proxies or not. The client can add the header in the original call and proxies just append their IPs into the list.

Reply:
Of course. But the point is, the first (as they prepend) can be trusted if we know there are proxies in our deployment. But definitely not when there aren't. That's why it has to be configurable.

Reply:
I deleted my previous comment, I wasn't thinking straight. :)

Reply:
No, the first can't be trusted, it's the one that can be provided by the original caller. Proxies _append_ into the end of the list.

```bash
$ curl -s -H "X-Forwarded-For: look_me_mama" https://httpbin.agw.arodevtest.hel.fi/anything?show_env | grep X-Forwarded-For
    "X-Forwarded-For": "look_me_mama,85.76.82.71:39599",
```

Reply:
The usage of the header should be configured because when there are no proxies in front, we shouldn't let the user  give the IP-address themselves using the header.

In proxy case, we should have a setting to denote the number of proxies we trust and then take the nth from the end. I think.

Reply:
The first one _could_ be made trustful, if the edge proxy (or API GW or whatever there is) would **replace** the X-Forwarded-For header, i.e. discard anything that it potentially received from the client. But it looks that at least the API GW tried above doesn't behave like that.

Reply:
The whole thing is kind of a mess. Because of course the user could use proxies even before the request hits our gateways et al. Then the first IP would be lost.

Reply:
I think the conclusion is that there is no 100% correct solution. Perhaps we should settle for a "mostly works" solution.

Reply:
I agree. That could be: take the first if from the header if the "USE_X_FORWARDED_FOR" (or something) setting is true, else use REMOTE_ADDR.